File: backend/api/models/client_credit.py
from datetime import date

from django.contrib.auth.models import User
from django.core.validators import MinValueValidator
from django.db import models


class AvoirClient(models.Model):
    """Note de crédit émise au client pour une vente non annulable."""

    class Statut(models.TextChoices):
        BROUILLON = 'BROUILLON', 'Brouillon'
        VALIDEE = 'VALIDEE', 'Validée'
        ANNULEE = 'ANNULEE', 'Annulée'

    class TypeMotif(models.TextChoices):
        ERREUR = 'ERREUR', 'Erreur'
        RETOUR = 'RETOUR', 'Retour'
        REMISE = 'REMISE', 'Remise'
        AUTRE = 'AUTRE', 'Autre'

    numero = models.CharField(max_length=50, unique=True, blank=True)
    facture_origine = models.ForeignKey(
        'Facture', on_delete=models.SET_NULL, null=True, blank=True,
        related_name='avoirs_clients',
    )
    client = models.ForeignKey(
        'Client', on_delete=models.SET_NULL, null=True, blank=True,
        related_name='avoirs_clients',
    )
    date = models.DateField(default=date.today)
    montant_total = models.DecimalField(max_digits=12, decimal_places=2)
    statut = models.CharField(max_length=10, choices=Statut.choices, default=Statut.BROUILLON)
    type_motif = models.CharField(max_length=10, choices=TypeMotif.choices, default=TypeMotif.AUTRE)
    created_by = models.ForeignKey(
        User, on_delete=models.SET_NULL, null=True, blank=True,
        related_name='avoirs_clients_created',
    )
    notes = models.TextField(blank=True, default='')

    class Meta:
        ordering = ['-date', '-id']
        verbose_name = 'Avoir client'
        verbose_name_plural = 'Avoirs clients'

    def __str__(self):
        return self.numero

    def save(self, *args, **kwargs):
        if not self.numero:
            self.numero = self.generate_numero()
        super().save(*args, **kwargs)

    @classmethod
    def generate_numero(cls):
        """Génère un numéro au format AVC-YYYYMM-XXXX."""
        today = date.today()
        prefix = f"AVC-{today.strftime('%Y%m')}"
        last = cls.objects.filter(numero__startswith=prefix).order_by('-numero').first()
        try:
            sequence = int(last.numero.rsplit('-', 1)[-1]) + 1 if last else 1
        except (TypeError, ValueError):
            sequence = 1
        return f'{prefix}-{sequence:04d}'


class LigneAvoirClient(models.Model):
    avoir_client = models.ForeignKey(AvoirClient, on_delete=models.CASCADE, related_name='lignes')
    produit = models.ForeignKey('Produit', on_delete=models.PROTECT)
    quantity = models.IntegerField(validators=[MinValueValidator(1)])
    prix_unitaire = models.DecimalField(max_digits=10, decimal_places=2)
    remise = models.DecimalField(max_digits=10, decimal_places=2, default=0)
    tva = models.DecimalField(max_digits=5, decimal_places=2)
    lot = models.CharField(max_length=100, blank=True, default='')
    stock_lot = models.ForeignKey('StockLot', on_delete=models.SET_NULL, null=True, blank=True)

    class Meta:
        verbose_name = "Ligne d'avoir client"
        verbose_name_plural = "Lignes d'avoir client"

    def __str__(self):
        return f'{self.avoir_client.numero} - {self.produit} x {self.quantity}'

File: frontend/frontend/src/types/clientCredit.ts
import type { ProduitModel } from './catalog';

export type ClientCreditStatus = 'BROUILLON' | 'VALIDEE' | 'ANNULEE';

export type ClientCreditTypeMotif = 'ERREUR' | 'RETOUR' | 'REMISE' | 'AUTRE';

export type RefundMethod = 'cash' | 'credit';

export interface ClientCreditLine {
    id: number;
    avoir_client: number;
    produit: number | ProduitModel;
    produit_nom?: string;
    quantity: number;
    prix_unitaire: string;
    remise: string;
    tva: string;
    lot?: string;
    stock_lot?: number | null;
}

export interface ClientCredit {
    id: number;
    numero: string;
    facture_origine: number | null;
    facture_numero?: string;
    client: number | null;
    client_name?: string;
    date: string;
    montant_total: string;
    statut: ClientCreditStatus;
    type_motif: ClientCreditTypeMotif;
    created_by?: number | null;
    created_by_name?: string;
    notes: string;
    lignes: ClientCreditLine[];
}

export interface ClientCreditFilters {
    search?: string;
    statut?: ClientCreditStatus;
    client?: number;
    facture_origine?: number;
    type_motif?: ClientCreditTypeMotif;
    date_debut?: string;
    date_fin?: string;
    page?: number;
    page_size?: number;
}

export interface ClientCreditCreatePayload {
    facture_origine?: number | null;
    client?: number | null;
    montant_total: string | number;
    type_motif: ClientCreditTypeMotif;
    notes?: string;
    validated_by_id?: number | null;
    sudo_password?: string;
    lignes: Array<{
        produit: number;
        quantity: number;
        prix_unitaire: string | number;
        remise?: string | number;
        tva?: string | number;
        lot?: string;
        stock_lot?: number | null;
    }>;
}

export interface ClientCreditValidatePayload {
    refund_method: RefundMethod;
    validated_by_id?: number | null;
    sudo_password?: string;
}

export interface InvoiceForCreditData {
    facture_origine: number;
    facture_numero: string;
    client: number | null;
    client_name: string;
    montant_total: string;
    lignes: Array<{
        produit: number;
        produit_nom: string;
        quantity: number;
        prix_unitaire: string;
        remise: string;
        tva: string;
        lot: string;
        stock_lot: number | null;
        date_expiration?: string | null;
        use_lot_management?: boolean;
    }>;
}
